File: 0688-Knight Probability in Chessboard/0688-Knight Probability in Chessboard.cpp
class Solution {
public:
    double knightProbability(int N, int K, int r, int c) {
        vector<vector<double>> dp(N, vector<double>(N));
        dp[r][c] = 1;
        vector<int> dx = {-2, -2, -1, -1, 1, 1, 2, 2};
        vector<int> dy = {-1, 1, -2, 2, -2, 2, -1, 1};
        for (int k = 0; k < K; ++k) {
            vector<vector<double>> temp(N, vector<double>(N));
            for (int i = 0; i < N; ++i) {
                for (int j = 0; j < N; ++j) {
                    for (int l = 0; l < dx.size(); ++l) {
                        int nx = i + dx[l];
                        int ny = j + dy[l];
                        if (nx >= 0 && nx < N && ny >= 0 && ny < N) {
                            temp[i][j] += dp[nx][ny] / 8;
                        }
                    }
                }
            }
            dp = temp;
        }
        
        double sum = 0;
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                sum += dp[i][j];
            }
        }
        
        return sum;
    }
};

File: 0601-0700/0688-Knight Probability in Chessboard/0688-Knight Probability in Chessboard.py
class Solution:
    def knightProbability(self, N: int, K: int, r: int, c: int) -> float:
        dp = [[0]*N for _ in range(N)]
        dp[r][c] = 1
        for _ in range(K):
            temp = [[0]*N for _ in range(N)]
            for i in range(N):
                for j in range(N):
                    for nr, nc in (i-2, j-1), (i-2, j+1), (i-1, j-2), (i-1, j+2), (i+1, j-2), (i+1, j+2), (i+2, j-1), (i+2, j+1):
                        if 0 <= nr < N and 0 <= nc < N:
                            temp[i][j] += dp[nr][nc] / 8
            dp = temp
        return sum(map(sum, dp))
